Microsoft.Toolkit.Uwp.UI.Controls/Carousel/CarouselPanel.cs

```diff
@@ -0,0 +1,455 @@
+﻿// ******************************************************************
+// Copyright (c) Microsoft. All rights reserved.
+// This code is licensed under the MIT License (MIT).
+// THE CODE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH
+// THE CODE OR THE USE OR OTHER DEALINGS IN THE CODE.
+// ******************************************************************
+
+using System;
+using Windows.Foundation;
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Controls;
+using Windows.UI.Xaml.Input;
+using Windows.UI.Xaml.Media;
+using Windows.UI.Xaml.Media.Animation;
+
+namespace Microsoft.Toolkit.Uwp.UI.Controls
+{
+    public class CarouselPanel : Panel
+    {
+        // prevent ArrangeOverride on items adding
+        private bool isUpdatingPosition;
+
+        // Storyboard on gesture
+        private Storyboard storyboard = new Storyboard();
+
+        // temp size
+        private double desiredWidth;
+        private double desiredHeight;
+
+        private Carousel carouselControl;
+
+        public CarouselPanel()
+        {
+            this.IsHitTestVisible = true;
+
+            this.ManipulationMode = ManipulationModes.All;
+            this.ManipulationCompleted += OnManipulationCompleted;
+            this.ManipulationDelta += OnManipulationDelta;
+            this.Tapped += OnTapped;
+        }
+
+        /// <summary>
+        /// Gets the Current Carousel control
+        /// </summary>
+        public Carousel Carousel
+        {
+            get { return carouselControl ?? (carouselControl = this.FindVisualAscendant<Carousel>()); }
+        }
+
+        /// <summary>
+        /// Tapp an item
+        /// </summary>
+        private void OnTapped(object sender, TappedRoutedEventArgs e)
+        {
+            var position = e.GetPosition(this);
+
+            for (int i = 0; i < this.Children.Count; i++)
```

Comment:
Maybe.
Don't remember exactly why I made it like this, but in my opinion:

1. I don't want CarouselItem, to let more flexibility to developers
2. I think I made it like that because of Virtualization (less events to subscribe)